Fails on Windows Agent

2017-11-23T18:23:51.3938674Z ##[section]Starting: helm dependency
2017-11-23T18:23:51.4378671Z ==============================================================================
2017-11-23T18:23:51.4378671Z Task         : Work with Helm packages
2017-11-23T18:23:51.4378671Z Description  : Build, install, and update helm packages.
2017-11-23T18:23:51.4378671Z Version      : 0.1.0
2017-11-23T18:23:51.4378671Z Author       : Don Ebben
2017-11-23T18:23:51.4378671Z Help         : [More Information](https://github.com/debben/vsts-helm-extension)
2017-11-23T18:23:51.4378671Z ==============================================================================
2017-11-23T18:23:52.7307830Z Downloading: https://storage.googleapis.com/kubernetes-helm/helm-v2.7.0-windows-amd64.zip
2017-11-23T18:23:52.9464515Z ##[error]Unexpected HTTP response: 404
2017-11-23T18:23:52.9634505Z ##[section]Finishing: helm dependency

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,0 +1,14 @@
+# Changelog
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
+
+## [Unreleased]
+
+## [0.1.4] - 2017-03-08
+### Added
+- A changelog!
+
+### Changed
+- Added --force-local option when using tar on Windows environments.

--- a/Helm/package.json
+++ b/Helm/package.json
@@ -1,5 +1,6 @@
 {
   "dependencies": {
+    "@types/node": "^8.0.54",
     "del": "2.2.0",
     "js-yaml": "3.6.1",
     "request-promise": "^4.2.1",

--- a/Helm/src/utilities.ts
+++ b/Helm/src/utilities.ts
@@ -3,9 +3,11 @@
 import * as fs from "fs";
 import * as path from "path";
 import * as tl from "vsts-task-lib/task";
+import * as trm from "vsts-task-lib/toolrunner";
 import * as toolLib from 'vsts-task-tool-lib/tool';
 import * as restm from 'typed-rest-client/RestClient';
 import * as os from "os";
+import * as uuidV4 from 'uuid/v4';
 
 //
 // Helm versions interface
@@ -105,7 +107,7 @@ async function acquireHelm(version: string): Promise<string> {
         }
 
         extPath = path.join(extPath, 'n'); // use as short a path as possible due to nested node_modules folders
-        extPath = await toolLib.extractTar(downloadPath);
+        extPath = await extractTar(downloadPath);
     }
     else {
         extPath = await toolLib.extractTar(downloadPath);
@@ -141,4 +143,61 @@ function ensureDirExists(dirPath : string) : void
     if (!fs.existsSync(dirPath)) {
         fs.mkdirSync(dirPath);
     }
+}
+
+/**
+ * The following code is used with attribution from 
+ * https://github.com/Microsoft/vsts-task-tool-lib and 
+ * caries with it the following copyright and MIT license
+ * statement:
+ * 
+ *  MIT License
+
+    Copyright (c) Microsoft Corporation. All rights reserved.
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE
+ * 
+ */
+
+function _getAgentTemp(): string {
+    tl.assertAgent('2.115.0');
+    let tempDirectory = tl.getVariable('Agent.TempDirectory');
+    if (!tempDirectory) {
+        throw new Error('Agent.TempDirectory is not set');
+    }
+
+    return tempDirectory;
+}
+
+async function extractTar(file: string): Promise<string> {
+
+    console.log(tl.loc('TOOL_LIB_ExtractingArchive'));
+    let dest = (() => {
+        let dest = path.join(_getAgentTemp(), uuidV4());
+    
+        tl.mkdirP(dest);
+        return dest;
+    })();
+
+    let tr: trm.ToolRunner = tl.tool('tar');
+    tr.arg(['xzC', dest.replace('\\','/'), '--force-local', '-f', file]);
+
+    await tr.exec();
+    return dest;
 }

--- a/Helm/src/utilities.ts
+++ b/Helm/src/utilities.ts
@@ -87,7 +87,7 @@ async function acquireHelm(version: string): Promise<string> {
     let fileName: string = osPlat == 'win32'? 'helm-v' + version + '-windows-' + arch :
                                                 'helm-v' + version + '-' + osPlat + '-' + arch;  
 
-    let urlFileName: string = fileName +  (osPlat == 'win32'? '.zip' : '.tar.gz');
+    let urlFileName: string = fileName +  '.tar.gz';
 
     let downloadUrl = 'https://storage.googleapis.com/kubernetes-helm/' + urlFileName;
 
@@ -105,7 +105,7 @@ async function acquireHelm(version: string): Promise<string> {
         }
 
         extPath = path.join(extPath, 'n'); // use as short a path as possible due to nested node_modules folders
-        extPath = await toolLib.extractZip(downloadPath);
+        extPath = await toolLib.extractTar(downloadPath);
     }
     else {
         extPath = await toolLib.extractTar(downloadPath);

--- a/vss-extension.json
+++ b/vss-extension.json
@@ -2,7 +2,7 @@
     "manifestVersion": 1,
     "id": "vsts-helm-extension",
     "name": "Helm Extension",
-    "version": "0.1.1",
+    "version": "0.1.2",
     "publisher": "debben",
     "targets": [{
         "id": "Microsoft.VisualStudio.Services"

--- a/vss-extension.json
+++ b/vss-extension.json
@@ -2,7 +2,7 @@
     "manifestVersion": 1,
     "id": "vsts-helm-extension",
     "name": "Helm Extension",
-    "version": "0.1.3",
+    "version": "0.1.4",
     "publisher": "debben",
     "targets": [{
         "id": "Microsoft.VisualStudio.Services"

--- a/vss-extension.json
+++ b/vss-extension.json
@@ -2,7 +2,7 @@
     "manifestVersion": 1,
     "id": "vsts-helm-extension",
     "name": "Helm Extension",
-    "version": "0.1.2",
+    "version": "0.1.3",
     "publisher": "debben",
     "targets": [{
         "id": "Microsoft.VisualStudio.Services"

